### PR TITLE
Test full QSL in setup_mock_browser

### PR DESCRIPTION
**Pull request by hemberger**
Related to #158.

Compare the expected and posted QSL as lists rather than sets
to avoid clobbering multiple elements with the same name.

For example:
```python
parse_qsl('x=1&x=1') != parse_qsl('x=1')
```
even though
```python
set(parse_qsl('x=1&x=1')) == set(parse_qsl('x=1'))
```
Since valid HTML can produce both query strings, we should be
able to distinguish between them in our tests.

### CODE REVIEW

File: tests/utils.py [hunk 1]
```diff
@@ -44,7 +44,7 @@ def setup_mock_browser(expected_post=None, text=choose_submit_form):
         def text_callback(request, context):
             # Python 2's parse_qsl doesn't like None argument
             query = parse_qsl(request.text) if request.text else ()
-            assert(set(query) == set(expected_post))
+            assert list(query) == list(expected_post)
```
Comment (moy) [inline tests/utils.py]: Do we need this `list(...)`? If so, this deserves a comment to explain why.

**Comment by moy** [comment]
OK, I think I understand the issue: we build `data` as a dictionary, which now is a dictionary of lists. Individual lists are ordered, but the dictionary isn't.

So, either we should

1) build data as a list of lists, or
2) use a comparison operator intermediate between "check the order" and "only check set": check number of repeated keys, but don't check the order of keys.

Option 1) might be better (in case some website relies on the order of keys), but AFAIK the order of fields in the data sent by browser is unspecified, so it's probably overkill to try to get the order right. So I'd lean towards 2), and this is rather simple to do: just sort the lists before comparison.

In short, I think we could write this instead:

    assert sorted(query) == sorted(expected_post)

File: tests/utils.py (hunk 1 above)
Comment (hemberger) [reply tests/utils.py]: The `list` is needed to ensure that `query` and `expected_post` are the same type. I think `query` is always a list, but `expected_post` was sometimes a tuple. I can add a comment, but we could also consider removing the cast for `query`.

**Comment by hemberger** [comment]
Yes, when checking the expected query string (at least with [TryIt](https://www.w3schools.com/html/tryit.asp?filename=tryhtml_form_submit)), the order was always the order of the input elements on the page. I don't know if this is set by any standard, but we're treading in the murky area of bad HTML design anyway, since repeated element names is rarely a good idea. :)

**Comment by moy** [comment]
The non-deterministic ordering we get is not for repeated elements, but for different keys i.e. generating `x=1&y=2` or `y=2&x=1`.

**Comment by hemberger** [comment]
I see, so for different keys, I can't imagine that the ordering matters (as you were saying), but if we use `sorted` to check the QSL, then won't that hide any errors with the ordering of elements with the same name?

**Comment by moy** [comment]
Then, we can use a sort only on the keys and keep the values ordered as they were:

```
def c(x, y):
    return cmp(x[0], y[0])
sorted(x, c)
```

**Comment by hemberger** [comment]
I think that if we refactor `_build_request` (something I was thinking about today, and may take a stab at tomorrow), then we can pass Requests a list of tuples for `data`. Looking at its source code, the list of tuples will create a deterministic query string:
```python
    def _encode_params(data):
        """Encode parameters in a piece of data.

        Will successfully encode parameters when passed as a dict or a list of
        2-tuples. Order is retained if data is a list of 2-tuples but arbitrary
        if parameters are supplied as a dict.
        """
```
Then we can do a simple equality operator and test that ordering is retained.

**Comment by moy** [comment]
Yes, that would work, but that just makes the code a bit heavier on our side:

First, you can't `data[key] = value` anymore. But a straightforward helper function should allow us to deal with a list of tuples while retaining a clean code on our side.

But more importantly, if we want the order to be the text order, we can't encode data as we do currently by iterating through elements types by types. Currently, we process all `input` tags, then all `textarea`, and then all `select`.

**Comment by moy** [comment]
Superseded by #219 .
